auth/api/iam/s2s_vptoken.go

```diff
@@ -143,7 +168,137 @@ func (r Wrapper) createAccessToken(issuer did.DID, issueTime time.Time, presenta
+			Code:        oauth.InvalidScope,
+			Description: fmt.Sprintf("unsupported scope for presentation exchange: %s", scope),
+		}
+	}
+
+	_, err := submission.Validate(*pexEnvelope, *definition)
+	if err != nil {
+		return nil, oauth.OAuth2Error{
+			Code:          oauth.InvalidRequest,
+			Description:   "presentation submission does not conform to Presentation Definition",
+			InternalError: err,
+		}
+	}
+	return definition, err
+}
+
+// validatePresentationValidity checks that the presentation is valid for a reasonable amount of time.
+func validatePresentationValidity(presentation vc.VerifiablePresentation) error {
+	created := credential.PresentationIssuanceDate(presentation)
+	expires := credential.PresentationExpirationDate(presentation)
+	if created == nil || expires == nil {
+		return oauth.OAuth2Error{
+			Code:        oauth.InvalidRequest,
+			Description: "presentation is missing creation or expiration date",
+		}
+	}
+	if expires.Sub(*created) > maxPresentationValidity {
+		return oauth.OAuth2Error{
+			Code:        oauth.InvalidRequest,
+			Description: fmt.Sprintf("presentation is valid for too long (max %s)", maxPresentationValidity),
+		}
+	}
+	return nil
+}
+
+// validatePresentationSigner checks if the presenter of the VP is the same as the subject of the VCs being presented.
+func validatePresentationSigner(presentation vc.VerifiablePresentation) error {
+	ok, err := credential.PresenterIsCredentialSubject(presentation)
+	if err != nil {
+		return oauth.OAuth2Error{
+			Code:        oauth.InvalidRequest,
+			Description: err.Error(),
+		}
+	}
+	if !ok {
+		return oauth.OAuth2Error{
+			Code:        oauth.InvalidRequest,
+			Description: "presentation signer is not credential subject",
+		}
+	}
+	return nil
+}
+
+// validatePresentationNonce checks if the nonce has been used before; 'jti' claim for JWTs or LDProof's 'nonce' for JSON-LD.
+func (r *Wrapper) validatePresentationNonce(presentation vc.VerifiablePresentation) error {
+	var nonce string
+	switch presentation.Format() {
+	case vc.JWTPresentationProofFormat:
+		nonce = presentation.JWT().JwtID()
+		if nonce == "" {
+			return oauth.OAuth2Error{
+				Code:        oauth.InvalidRequest,
+				Description: "presentation is missing jti",
+			}
+		}
+	case vc.JSONLDPresentationProofFormat:
+		proof, err := credential.ParseLDProof(presentation)
+		if err != nil || proof.Nonce == nil {
+			return oauth.OAuth2Error{
+				Code:          oauth.InvalidRequest,
+				InternalError: err,
+				Description:   "presentation has invalid proof or nonce",
+			}
+		}
+		nonce = *proof.Nonce
+	}
+
+	nonceStore := r.storageEngine.GetSessionDatabase().GetStore(maxPresentationValidity, "s2s", "nonce")
+	err := nonceStore.Get(nonce, new(bool))
+	if !errors.Is(err, storage.ErrNotFound) {
+		if err != nil {
+			// unable to check nonce
+			return err
+		}
+		return oauth.OAuth2Error{
+			Code:        oauth.InvalidRequest,
+			Description: "presentation nonce has already been used",
+		}
+	}
```

Comment:
It's right, but confusing since good cases should return `storage.ErrNotFound`. So I rewrote it a bit to make it more explicit. I also decided to always write the nonce as used, even if the an error occurs while checking it. So a VP can really only be used once.